src/utils/deploy.rs

```diff
@@ -0,0 +1,169 @@
+// SPDX-FileCopyrightText: 2020 Serokell <https://serokell.io/>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+use std::borrow::Cow;
+use tokio::process::Command;
+
+fn build_activate_command(
+    activate_path_str: String,
+    sudo: &Option<String>,
+    profile_path: &str,
+    closure: &str,
+    bootstrap_cmd: &Option<String>,
+    auto_rollback: bool,
+    temp_path: &Cow<str>,
+    confirm_timeout: u16,
+    magic_rollback: bool,
+) -> String {
+    let mut self_activate_command = format!(
+        "{} '{}' '{}' --temp-path {} --confirm-timeout {}",
+        activate_path_str, profile_path, closure, temp_path, confirm_timeout
+    );
+
+    if magic_rollback {
+        self_activate_command = format!("{} --magic-rollback", self_activate_command);
+    }
+
+    if auto_rollback {
+        self_activate_command = format!("{} --auto-rollback", self_activate_command);
+    }
+
+    if let Some(ref bootstrap_cmd) = bootstrap_cmd {
+        self_activate_command = format!(
+            "{} --bootstrap-cmd '{}'",
+            self_activate_command, bootstrap_cmd
+        );
+    }
+
+    if let Some(sudo_cmd) = &sudo {
+        self_activate_command = format!("{} {}", sudo_cmd, self_activate_command);
+    }
+
+    self_activate_command
+}
+
+#[test]
+fn test_activation_command_builder() {
+    let activate_path_str = "/blah/bin/activate".to_string();
+    let sudo = Some("sudo -u test".to_string());
+    let profile_path = "/blah/profiles/test";
+    let closure = "/blah/etc";
+    let bootstrap_cmd = None;
+    let auto_rollback = true;
+    let temp_path = &"/tmp/deploy-rs".into();
+    let confirm_timeout = 30;
+    let magic_rollback = true;
+
+    assert_eq!(
+        build_activate_command(
+            activate_path_str,
+            &sudo,
+            profile_path,
+            closure,
+            &bootstrap_cmd,
+            auto_rollback,
+            temp_path,
+            confirm_timeout,
+            magic_rollback
+        ),
+        "sudo -u test /blah/bin/activate '/blah/profiles/test' '/blah/etc' --temp-path /tmp/deploy-rs --confirm-timeout 30 --magic-rollback --auto-rollback"
+            .to_string(),
+    );
+}
+
+pub async fn deploy_profile(
+    deploy_data: &super::DeployData<'_>,
+    deploy_defs: &super::DeployDefs<'_>,
+) -> Result<(), Box<dyn std::error::Error>> {
+    info!(
+        "Activating profile `{}` for node `{}`",
+        deploy_data.profile_name, deploy_data.node_name
+    );
+
+    let activate_path_str = super::deploy_path_to_activate_path_str(&deploy_defs.current_exe)?;
+
+    let temp_path: Cow<str> = match &deploy_data.merged_settings.temp_path {
+        Some(x) => x.into(),
+        None => "/tmp/deploy-rs".into(),
+    };
```

Comment:
```suggestion
    let temp_path: Cow<str> = deploy_data.merged_settings.temp_path.as_ref().unwrap_or("/tmp/deploy-rs").into();
```

(Untested, don't just commit it in)

Reply:
`.as_deref()` instead, and some other tweaks, but it always seems to produce totally different assembly when I test it in isolated cases, and I don't imagine it's any more efficient so I think it's better to stick with `match` for this kind of thing